### ISSUE TITLE
Inconsistent model description and pre-trained model

### ISSUE THREAD

**Issue by ljk628** [facebook/fb.resnet.torch#49]
The pretrained models (ResNet-18/34/50) have no BN layer in the shortcut connection where the number of feature maps changes. which is inconsistent with the model description [here](https://github.com/facebook/fb.resnet.torch/blob/b7e6724e8a079fdce1de2ea8dad876e18183f911/models/resnet.lua#L34). 

shortcut in pretrained model:

      (1): nn.ConcatTable {
        input
          |`-> (1): nn.Sequential {
          |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> output]
          |      (1): cudnn.SpatialConvolution(64 -> 128, 3x3, 2,2, 1,1)
          |      (2): nn.SpatialBatchNormalization
          |      (3): cudnn.ReLU
          |      (4): cudnn.SpatialConvolution(128 -> 128, 3x3, 1,1, 1,1)
          |      (5): nn.SpatialBatchNormalization
          |    }
          |`-> (2): cudnn.SpatialConvolution(64 -> 128, 1x1, 2,2)
           ... -> output
      }

shortcut in model definition:

      (1): nn.ConcatTable {
        input
          |`-> (1): nn.Sequential {
          |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> output]
          |      (1): cudnn.SpatialConvolution(64 -> 128, 3x3, 2,2, 1,1) without bias
          |      (2): nn.SpatialBatchNormalization
          |      (3): cudnn.ReLU
          |      (4): cudnn.SpatialConvolution(128 -> 128, 3x3, 1,1, 1,1) without bias
          |      (5): nn.SpatialBatchNormalization
          |    }
          |`-> (2): nn.Sequential {
          |      [input -> (1) -> (2) -> output]
          |      (1): cudnn.SpatialConvolution(64 -> 128, 1x1, 2,2) without bias
          |      (2): nn.SpatialBatchNormalization
          |    }
           ... -> output

**Comment by loeweX** [comment]
In the resnet-50 model, the BN has even moved behind the addition of the two blocks:

`    
(3): nn.Sequential {
      [input -> (1) -> (2) -> (3) -> (4) -> output]
      (1): nn.ConcatTable {
        input
          |`-> (1): nn.Sequential {
          |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> output]
          |      (1): cudnn.SpatialConvolution(2048 -> 512, 1x1)
          |      (2): nn.SpatialBatchNormalization
          |      (3): cudnn.ReLU
          |      (4): cudnn.SpatialConvolution(512 -> 512, 3x3, 1,1, 1,1)
          |      (5): nn.SpatialBatchNormalization
          |      (6): cudnn.ReLU
          |      (7): cudnn.SpatialConvolution(512 -> 2048, 1x1)
          |    }
          |`-> (2): nn.Identity
           ... -> output
      }
      (2): nn.CAddTable
     ** (3): nn.SpatialBatchNormalization**
      (4): cudnn.ReLU
    }
` 
Although in this blog post (http://torch.ch/blog/2016/02/04/resnets.html) it was stated that 'putting batch normalization after the addition significantly hurts test error on CIFAR'.

**Comment by loeweX** [comment]
In the resnet-50 model the BN has even moved behind the addition of the two blocks:

```
    (3): nn.Sequential {
      [input -> (1) -> (2) -> (3) -> (4) -> output]
      (1): nn.ConcatTable {
        input
          |`-> (1): nn.Sequential {
          |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> output]
          |      (1): cudnn.SpatialConvolution(2048 -> 512, 1x1)
          |      (2): nn.SpatialBatchNormalization
          |      (3): cudnn.ReLU
          |      (4): cudnn.SpatialConvolution(512 -> 512, 3x3, 1,1, 1,1)
          |      (5): nn.SpatialBatchNormalization
          |      (6): cudnn.ReLU
          |      (7): cudnn.SpatialConvolution(512 -> 2048, 1x1)
          |    }
          |`-> (2): nn.Identity
           ... -> output
      }
      (2): nn.CAddTable
      (3): nn.SpatialBatchNormalization
      (4): cudnn.ReLU
    }
```

Although in this blog post (http://torch.ch/blog/2016/02/04/resnets.html) it was stated that 'putting batch normalization after the addition significantly hurts test error on CIFAR'.

**Comment by colesbury** [comment]
@ljk628, I misunderstood the the placement of the batch norm layers when I trained those models. I don't think it makes much of a difference in terms of accuracy.

@loeweX, that's noted on the https://github.com/facebook/fb.resnet.torch/tree/master/pretrained. For some reason, those results on CIFAR didn't transfer to ImageNet.

At some point, I'd like to retrain all the models with the latest definition for consistency.

**Comment by ljk628** [comment]
Thanks! @colesbury I think the pretrained model is probably closer to the original paper. The new preresnet [model](https://github.com/facebook/fb.resnet.torch/blob/65e675a8cc7c96918d140656af55941afbfa8104/models/resnet.lua#L34) and Kaiming's [implementation](https://github.com/KaimingHe/resnet-1k-layers/blob/fde7ee93c464d8312ca9b3df1c488cb245a9e2c1/resnet-pre-act.lua#L83) have no BN layer either in the short cut connection. So I guess it would be it easier to remove this [line](https://github.com/facebook/fb.resnet.torch/blob/65e675a8cc7c96918d140656af55941afbfa8104/models/resnet.lua#L34) and the sequential wrapper from resnet.lua :)

**Comment by 1adrianb** [comment]
The same thing is true for the ResNet-200 model and the latest trained model is not identical with the one from preresnet.lua
